crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -3193,10 +3193,7 @@ impl<'db> TypeInferenceBuilder<'db> {
             (UnaryOp::Not, ty) => ty.bool(self.db()).negate().into_type(self.db()),
             (_, Type::Any) => Type::Any,
             (_, Type::Unknown) => Type::Unknown,
-            (
-                op @ (UnaryOp::UAdd | UnaryOp::USub | UnaryOp::Invert),
-                Type::Instance(InstanceType { class }),
-            ) => {
+            (op @ (UnaryOp::UAdd | UnaryOp::USub | UnaryOp::Invert), _) => {
```

Comment:
I'm not so sure explicitly listing all variants is always the best answer. In this case, the handling in this arm really is a generic implementation of the semantics in terms of other type semantics. It should always be correct for _any_ type (even the ones we special case above here), assuming a correct implementation of calling dunder methods (`to_meta_type`, `member`, etc -- all of which _do_ use exhaustive matches and we'll be forced to consider for any new Type variant.) Thus, a special case for a new Type variant here would not be required for correctness, it would just potentially offer more precision. I kind of think it's fine, possibly even good, for that "more precision" to be driven only by actual user needs, and not something we are forced to consider.

But I'm really fine with it either way.

Reply:
Yes, you're right that there isn't a correctness issue here. But I would expect more precision here to be both trivial to accomplish for any new variants and simpler for us to compute. The `IntLiteral()` branches above really involve much less work for us than doing the full lookup for an instance type's dunder method in typeshed or user code (which could involve materialising an MRO!). I don't really see a reason _not_ for us to infer the more precise type whenever possible if it's trivial to do so

Reply:
Sounds like there's a -0 vote from Carl and a +1 from Alex, so I went ahead and expanded the clauses out.  We can always put it back to `_` in the future if we find this more unwieldy than the benefit.